File: apps/web/src/api/roleSkillSources.ts
import { get } from './client';
import { parseFrameworkSource, type FrameworkSourceView } from './frameworkSource';
import { isSkillObservation, type SkillObservation } from './skillInventory';

interface SourceRoot { kind: 'workspace_skills' | 'project_agent_skills' | 'profile_skills'; locator_sha256: string }
export interface RoleSkillSourcePage {
  schema_version: 'enterprise-role-skill-sources-view/v1' | 'enterprise-role-skill-sources-view/v2'; asset_id: string;
  status: 'source_unavailable' | 'historical_comparison'; framework_source: FrameworkSourceView;
  declared_roots: null | { schema_version: 'enterprise-role-skill-roots/v1';
    basis: 'agent_workspace' | 'default_workspace'; status: 'declared'; roots: SourceRoot[] }
    | { schema_version: 'enterprise-role-skill-roots/v2'; basis: 'hermes_profile_layout'; status: 'layout_candidate'; roots: SourceRoot[] };
  coverage: 'page_of_device_installations'; next_cursor: string | null;
  runtime_status: 'unverified'; effective_permissions: null;
  items: { installation_id: string; locator_sha256: string;
    relationship_status: 'historical_source_match' | 'outside_declared_sources' | 'unresolved';
    matched_sources: SourceRoot[]; observation: SkillObservation | null }[];
}
const object = (v: unknown): v is Record<string, unknown> => !!v && typeof v === 'object' && !Array.isArray(v);
const exact = (v: Record<string, unknown>, keys: string[]) => Object.keys(v).length === keys.length && keys.every(k => Object.hasOwn(v, k));
const digest = (v: unknown): v is string => typeof v === 'string' && /^[a-f0-9]{64}$/.test(v);
const installationId = (v: unknown): v is string => typeof v === 'string' && /^ski_[A-Za-z0-9_-]{1,60}$/.test(v);
function root(v: unknown): v is SourceRoot {
  return object(v) && exact(v, ['kind', 'locator_sha256'])
    && (v.kind === 'workspace_skills' || v.kind === 'project_agent_skills' || v.kind === 'profile_skills') && digest(v.locator_sha256);
}

export function parseRoleSkillSources(value: unknown, assetId: string, cursor?: string): RoleSkillSourcePage {
  const fail = () => { throw new Error('技能来源响应无法核验'); };
  if (!object(value) || !exact(value, ['schema_version', 'asset_id', 'status', 'framework_source', 'declared_roots', 'coverage', 'items', 'next_cursor', 'runtime_status', 'effective_permissions'])
    || (typeof value.schema_version !== 'string' || !['enterprise-role-skill-sources-view/v1', 'enterprise-role-skill-sources-view/v2'].includes(value.schema_version)) || value.asset_id !== assetId
    || value.coverage !== 'page_of_device_installations' || value.runtime_status !== 'unverified'
    || value.effective_permissions !== null || !Array.isArray(value.items) || value.items.length > 50) return fail();
  const source = parseFrameworkSource(value.framework_source, assetId);
  const hermes = value.schema_version === 'enterprise-role-skill-sources-view/v2';
  if (source.schema_version !== `enterprise-framework-source-view/${hermes ? 'v2' : 'v1'}`) return fail();
  if (value.status === 'source_unavailable') {
    if (value.declared_roots !== null || value.items.length !== 0 || value.next_cursor !== null) return fail();
    return value as unknown as RoleSkillSourcePage;
  }
  const declared = value.declared_roots;
  if (value.status !== 'historical_comparison' || source.status !== 'historical_reported_source'
    || !object(declared) || !exact(declared, ['schema_version', 'basis', 'status', 'roots'])
    || !Array.isArray(declared.roots) || !declared.roots.every(root)) return fail();
  if (hermes) {
    if (declared.schema_version !== 'enterprise-role-skill-roots/v2' || declared.status !== 'layout_candidate'
      || declared.basis !== 'hermes_profile_layout' || declared.roots.length !== 1
      || declared.roots[0].kind !== 'profile_skills') return fail();
  } else if (declared.schema_version !== 'enterprise-role-skill-roots/v1' || declared.status !== 'declared'
    || (typeof declared.basis !== 'string' || !['agent_workspace', 'default_workspace'].includes(declared.basis)) || declared.roots.length !== 2
    || declared.roots[0].kind !== 'workspace_skills' || declared.roots[1].kind !== 'project_agent_skills'
    || declared.roots[0].locator_sha256 === declared.roots[1].locator_sha256) return fail();
  const roots = declared.roots as SourceRoot[];
  let previous = cursor ?? '';
  for (const item of value.items) {
    if (!object(item) || !exact(item, ['installation_id', 'locator_sha256', 'relationship_status', 'matched_sources', 'observation'])
      || !installationId(item.installation_id) || item.installation_id <= previous || !digest(item.locator_sha256)
      || !Array.isArray(item.matched_sources) || item.matched_sources.length > 2 || !item.matched_sources.every(root)) return fail();
    previous = item.installation_id;
    const matches = item.matched_sources as SourceRoot[];
    if (new Set(matches.map(r => r.kind)).size !== matches.length
      || !matches.every(r => roots.some(d => d.kind === r.kind && d.locator_sha256 === r.locator_sha256))) return fail();
    if (item.relationship_status === 'unresolved') {
      if (matches.length || item.observation !== null) return fail();
    } else {
      if ((typeof item.relationship_status !== 'string' || !['historical_source_match', 'outside_declared_sources'].includes(item.relationship_status))
        || (item.relationship_status === 'historical_source_match') !== (matches.length > 0)
        || !object(item.observation) || !exact(item.observation, ['observation_id', 'manifest_sha256', 'parser_version', 'parse_status', 'name', 'allowed_tools_present', 'declared_tools', 'observed_at', 'batch_digest'])
        || !isSkillObservation(item.observation)) return fail();
    }
  }
  if (value.next_cursor !== null && (!installationId(value.next_cursor) || !value.items.length || value.next_cursor !== previous)) return fail();
  return value as unknown as RoleSkillSourcePage;
}

export async function getRoleSkillSources(assetId: string, cursor?: string): Promise<RoleSkillSourcePage> {
  if (!/^[A-Za-z0-9][A-Za-z0-9_.:-]{0,63}$/.test(assetId) || (cursor !== undefined && !installationId(cursor))) throw new Error('来源查询标识无效');
  return parseRoleSkillSources(await get<unknown>(`/agents/${encodeURIComponent(assetId)}/skill-installation-sources`,
    { query: { cursor, limit: 50 } }), assetId, cursor);
}

File: apps/control-api/app/routers/role_skill_sources.py
"""Historical, signed location comparison; no runtime or permission inference."""

import hashlib
import json
from datetime import UTC

from fastapi import APIRouter, Depends, HTTPException, Query, Response
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.db import get_session
from app.evidence_signing import canonical_json, verify_hex_signature
from app.framework_source import unique_fields
from app.framework_source_view import project_framework_sources
from app.models import (
    AgentAsset,
    EdgeAgent,
    EdgeTask,
    RoleConfigurationObservation,
    SkillInstallation,
    SkillUploadReceipt,
)
from app.role_skill_roots import parse_role_skill_roots
from app.routers.role_configuration_history import configuration_history_query, project_configuration_snapshot
from app.routers.skill_inventory import inventory_query, project_observation
from app.security import Identity, ensure_permission, get_identity
from app.skill_upload import SkillBatchIn

router = APIRouter(tags=['inventory'])


def verified_batch(receipt, task, edge, tenant_id):
    try:
        signed = receipt.signed_payload.encode('utf-8')
        if (len(signed) > 2 * 1024 * 1024 or receipt.tenant_id != tenant_id
                or receipt.edge_agent_id != edge.id or task.environment_id != edge.environment_id
                or task.task_type != 'skill_scan' or task.result_digest != receipt.batch_digest
                or hashlib.sha256(signed).hexdigest() != receipt.batch_digest
                or not verify_hex_signature(edge.public_key_pem, signed, receipt.signature)):
            return None
        raw = json.loads(signed, object_pairs_hook=unique_fields)
        if not isinstance(raw, dict) or 'signature' in raw:
            return None
        body = SkillBatchIn.model_validate({**raw, 'signature': receipt.signature})
        payload = task.payload
        if (body.task_id != receipt.task_id or task.id != receipt.task_id
                or payload.get('target_device_identity') != edge.device_identity
                or payload.get('connector') != 'directory' or payload.get('inventory_kind') != 'skills'
                or not isinstance(payload.get('scope'), dict)
                or hashlib.sha256(canonical_json(payload['scope'])).hexdigest() != body.scope_digest):
            return None
        return {item.locator_sha256: item for item in body.observations}
    except (ValueError, TypeError, AttributeError, RecursionError):
        return None


def matched_observation(item, observation):
    return (observation is not None and item.manifest_sha256 == observation.manifest_sha256
            and item.parser_version == observation.parser_version and item.parse_status == observation.parse_status
            and item.name == observation.name and item.allowed_tools_present == observation.allowed_tools_present
            and item.declared_tools == observation.declared_tools
            and item.observed_at.astimezone(UTC).replace(tzinfo=None) == observation.observed_at)


@router.get('/api/v1/agents/{asset_id}/skill-installation-sources')
def role_skill_sources(
    asset_id: str, response: Response,
    cursor: str | None = Query(default=None, pattern=r'^ski_[A-Za-z0-9_-]{1,60}$'),
    limit: int = Query(default=50, ge=1, le=100),
    identity: Identity = Depends(get_identity), session: Session = Depends(get_session),
):
    asset = session.scalar(select(AgentAsset).where(
        AgentAsset.id == asset_id, AgentAsset.tenant_id == identity.tenant_id))
    if asset is None:
        raise HTTPException(404, 'not_found')
    ensure_permission(identity, 'agent:read')
    ensure_permission(identity, 'env:read')
    response.headers['Cache-Control'] = 'no-store'
    source = project_framework_sources(session, identity.tenant_id, [asset])[asset.id]
    hermes = asset.framework == 'hermes'
    version = 'v2' if hermes else 'v1'
    result = {'schema_version': f'enterprise-role-skill-sources-view/{version}', 'asset_id': asset.id,
              'status': 'source_unavailable', 'framework_source': source, 'declared_roots': None,
              'coverage': 'page_of_device_installations', 'items': [], 'next_cursor': None,
              'runtime_status': 'unverified', 'effective_permissions': None}
    try:
        roots = parse_role_skill_roots((asset.attributes or {}).get('skill_source_roots'))
    except (ValueError, TypeError, AttributeError, RecursionError):
        return result
    if (source['status'] != 'historical_reported_source'
            or roots['status'] != ('layout_candidate' if hermes else 'declared')
            or roots['schema_version'] != f'enterprise-role-skill-roots/{version}'
            or source['schema_version'] != f'enterprise-framework-source-view/{version}'
            or source['source']['framework'] != ('hermes' if hermes else 'openclaw')):
        return result
    result.update(status='historical_comparison', declared_roots=roots)
    result.update(installation_source_page(session, identity, source['source']['device_id'], roots, cursor, limit))
    return result


def installation_source_page(session, identity, device_id, roots, cursor, limit):
    result = {'items': [], 'next_cursor': None}
    query = inventory_query(identity.tenant_id).where(EdgeAgent.id == device_id)
    if cursor is not None:
        query = query.where(SkillInstallation.id > cursor)
    rows = session.execute(query.order_by(SkillInstallation.id).limit(limit + 1)).all()
    page = rows[:limit]
    if len(rows) > limit:
        result['next_cursor'] = page[-1][0].id
    digests = {row[3].batch_digest for row in page if row[3] is not None}
    receipts = {}
    receipt_rows = session.execute(select(SkillUploadReceipt, EdgeTask).join(
        EdgeTask, EdgeTask.id == SkillUploadReceipt.task_id).where(
            SkillUploadReceipt.tenant_id == identity.tenant_id,
            SkillUploadReceipt.edge_agent_id == device_id,
            SkillUploadReceipt.batch_digest.in_(digests),
    ).limit(101)).all()
    # More than one receipt per possible page digest is corrupt/ambiguous.
    # Never accept the first slice of an unbounded result as uniquely verified.
    for receipt, task in receipt_rows if len(receipt_rows) <= 100 else []:
        receipts.setdefault(receipt.batch_digest, []).append((receipt, task))
    checked = {}
    for installation, edge, _, observation in page:
        projected = {'installation_id': installation.id, 'locator_sha256': installation.locator_sha256,
                     'relationship_status': 'unresolved', 'matched_sources': [], 'observation': None}
        result['items'].append(projected)
        if observation is None:
            continue
        digest = observation.batch_digest
        entries = receipts.get(digest, [])
        if len(entries) != 1:
            continue
        receipt, task = entries[0]
        if digest not in checked:
            checked[digest] = verified_batch(receipt, task, edge, identity.tenant_id)
        batch = checked[digest]
        item = batch.get(installation.locator_sha256) if batch is not None else None
        if (item is None or not item.ancestor_sha256 or observation.batch_signature != receipt.signature
                or not matched_observation(item, observation)):
            continue
        matches = [root for root in roots['roots'] if root['locator_sha256'] in item.ancestor_sha256]
        projected.update(relationship_status='historical_source_match' if matches else 'outside_declared_sources',
                         matched_sources=matches, observation=project_observation(observation))
    return result


@router.get('/api/v1/agents/{asset_id}/configuration-observations/{observation_id}/skill-installation-sources')
def snapshot_skill_sources(
    asset_id: str, observation_id: str, response: Response,
    cursor: str | None = Query(default=None, pattern=r'^ski_[A-Za-z0-9_-]{1,60}$'),
    limit: int = Query(default=50, ge=1, le=100),
    identity: Identity = Depends(get_identity), session: Session = Depends(get_session),
):
    asset = session.scalar(select(AgentAsset).where(
        AgentAsset.id == asset_id, AgentAsset.tenant_id == identity.tenant_id))
    if asset is None:
        raise HTTPException(404, 'not_found')
    row = session.execute(configuration_history_query(identity.tenant_id, asset_id).where(
        RoleConfigurationObservation.id == observation_id)).first()
    if row is None:
        raise HTTPException(404, 'not_found')
    ensure_permission(identity, 'agent:read')
    ensure_permission(identity, 'env:read')
    response.headers['Cache-Control'] = 'no-store'
    snapshot = project_configuration_snapshot(row, asset.framework)
    hermes = asset.framework == 'hermes'
    version = 'v2' if hermes else 'v1'
    result = {'schema_version': f'enterprise-role-skill-snapshot-comparison/{version}', 'asset_id': asset_id,
              'configuration_observation': snapshot, 'status': 'snapshot_unavailable',
              'comparison_basis': 'latest_skill_observations_against_saved_configuration',
              'coverage': 'page_of_device_installations', 'items': [], 'next_cursor': None,
              'runtime_status': 'unverified', 'effective_permissions': None}
    if snapshot['status'] != 'recorded_snapshot':
        return result
    roots = snapshot['configuration']['skill_source_roots']
    if roots is None or roots['status'] != ('layout_candidate' if hermes else 'declared'):
        return result
    result['status'] = 'historical_comparison'
    result.update(installation_source_page(session, identity, snapshot['device_id'], roots, cursor, limit))
    return result
